core/src/feature.rs

```diff
@@ -10,14 +10,14 @@ use std::ffi::{c_void, CStr};
 /// A native plugin written in C would discover a host's features by iterating through an array of URIs and pointers. When it finds the URI of the feature it is looking for, it casts the pointer to the type of the feature interface and uses the information from the interface.
 ///
 /// In Rust, most of this behaviour is done internally and instead of simply casting a pointer, a safe feature descriptor, which implements this trait, is constructed using the [`from_raw_data`](#tymethod.from_raw_data) method.
-pub trait Feature: UriBound + Sized {
+pub trait Feature<'a>: UriBound + Sized {
```

Comment:
I am kinda hijacking this PR to talk about the recent `Features` changes you did, since this PR adds lifetimes to them.
As I told you in my email, I did not really understand how to implement a Feature with the changes, but now I understand that you change the `Feature` trait to not represent a LV2 Feature struct, but a pointer to an LV2 Feature struct instead, if I'm not mistaken. What pushed you to make this change?

The issue I have is that this API change makes it impossible to create LV2 Features without dropping to down to the `-sys` crate and to usage of `unsafe`, which is obnoxious for writing tests, and a straight-up no-go for writing LV2 Hosts. With the old API, you could say call `urid::Unap::new(...)` and call it a day, all safely.

Reply:
My main reason for this change is that you can't implement methods, for example safe wrappers for unsafe behavior, for types from other crates. Instead, the `Feature` trait is only implemented for structs that wrap and abstract the raw feature descriptors in a safe way, as it is done with the `map` and `unmap` feature. This way, a plugin programmer only needs to fetch the feature from the feature container and use it right away. I think this is quite nice! I did add the lifetime since I want to use `'static` as least as possible. In use, the lifetime `'a` will probably be resolved to `'static`, but you'll never know! 

Of course, it is impossible to create LV2 features without dropping down to the `-sys` crate! :wink: Features are functionalities that hosts provide and since we interact with the host via a C API, we have to use the `-sys` crates. Looking the other way around, if we are going to create a LV2 host in Rust, we also have to provide all of these features via structs from a `-sys` crate, constantly violating the semantic rules of Rust. Yes, that makes creating tests harder, but that's just how it is, I'm sorry about that!

Reply:
Oh, actually I meant that the `Features` used to represent *newtypes* around the `_sys` feature struct, and now they are newtypes around a pointer to the `_sys` feature struct. That's the change I was asking about, since in both cases you can add safe wrappers and/or helpers on those structs. :slightly_smiling_face:  (See the `Map` feature I had implemented on my prototype for example)

Reply:
The host provides only a pointer to the raw feature and since features aren't necessarily clonable, we may have to work with references instead. Also, as you can see with the marker features, the safe feature structs don't need to have that lifetime too.

If a safe feature struct has to hold a reference to the raw feature, it can do that and the lifetime of the trait helps with that. But the opposite is also possible. In this case the lifetime of the trait is only the lifetime of the raw feature reference.

Reply:
… I am not sure to quite comprehend what you're describing, I'm sorry xD
I think I'm going to illustrate it with code, it'll be simpler!

Basically, what we had was like this:

```rust
// Feature declaration
#[repr(C)]
pub struct MyFeature {
  inner: sys::LV2_MyFeature
  // Possibly a PhantomData here if we need some extra lifetime for safety reasons
}

// Usage (plugin developer)
#[derive(Features)]
pub struct PluginFeatures {
  my_feature: MyFeature
}
```

However, now I understand features cannot necessarily be copied/cloned (and I guess LV2 expects us to hold a pointer to it, so I guess it may not be even correct to do that). So you changed to this:

```rust
// Feature declaration
pub struct MyFeature<'a> {
  inner: &'a sys::LV2_MyFeature
}

// Usage (plugin developer)
#[derive(Features)]
pub struct PluginFeatures {
  my_feature: MyFeature
}
```

That fixes the issue, however we can't do things like this anymore:

```rust
impl<'a> MyFeature<'a> {
  /// Used by hosts to instanciate the feature, abstracting over the inner struct
  pub fn new(some_param: Foo) -> Self {
    Self { inner: sys::LV2_MyFeature { /* put the raw function pointers, allocate some handle perhaps, and so on */ } }
    // Here we create a new value, but the struct expects a reference, but creating it here would make it dangling!
  }
}
```

To create an abstraction for this, we would need to basically create a second struct here that would not use a reference, only for the host-side, which would be quite annoying I think.

My proposal is to do this instead:

```rust
// Feature declaration
#[repr(C)]
pub struct MyFeature {
  inner: sys::LV2_MyFeature
  // Possibly a PhantomData here if we need some extra lifetime for safety reasons
}

// Usage (plugin developer)
#[derive(Features)]
pub struct PluginFeatures<'a> {
  my_feature: &'a MyFeature // Now uses an explicit reference
}

// Now, we can do this again!
impl<'a> MyFeature<'a> {
  /// Used by hosts to instanciate the feature, abstracting over the inner struct
  pub fn new(some_param: Foo) -> Self {
    Self { inner: sys::LV2_MyFeature { /* put the raw function pointers, allocate some handle perhaps, and so on */ } }
  }
}
```

This is slightly more verbose on the plugin side since the references and their lifetimes are explicit now (I don't think it's too bad however), but now we can now properly instantiate the features again, allowing hosts to make use of the library.

There is a side-effect though, it's the fact that now all features are behind immutable references, so you can't have some feature methods that need an `&mut self`, unless they're only meant to be called by the host.

However I think it is incorrect for plugins to have access to features mutably (even if the LV2 header takes a mutable pointer): beyond mutability, `&mut` in Rust actually describes an unique reference (we can still mutate with `&` through interior mutability). But it goes through the window as feature pointers are created and duplicated for each running instance. Using `&mut` pointers is basically instant UB here in this scenario.

Moreover, it forces the host developers to recognize the issue instantly (as they're only given a `&self` to work with). For instance with `urid::Map`, the underlying storage has to be put inside a `RefCell` for it to compile, or in a `Mutex` if the host is multi-threaded, which is the correct behavior as there should be no mutable concurrent accesses whatsoever.

Reply:
Are we actually discussing how a plugin should interact with host features or is this a general discussion on how plugin and host code should relate to another?

As far as I can tell, you want the host and the plugin to use the same structs for the same features. However, I don't think this is useful: The host is focused on providing the interface to the plugin and the plugin is focused on using it. These are completely different tasks that require their own, specialized types and code. Trying to fit it all into one piece of code will probably not work well.

For example, if I were to implement the mapping feature for a host, I would use something along the lines of this this:

``` Rust
use std::collections::HashMap;
use std::ffi::{c_void, CStr};
use std::mem::transmute;
use std::sync::Mutex;

pub struct Map {
    storage: Box<Mutex<HashMap<&'static CStr, URID>>>,
}

impl Map {
    unsafe extern "C" fn extern_map(handle: *mut c_void, uri: *const i8) -> u32 {
        let handle = (handle as *mut Self).as_ref().unwrap();
        let uri = CStr::from_ptr(uri);
        *handle.map(uri)
    }

    pub fn map(&self, uri: &CStr) -> URID {
        // Map the uri
    }

    pub fn create_interface(&self) -> sys::LV2_URID_Map {
        sys::LV2_URID_Map {
            handle: unsafe {
                // I know, you should not do that, but I didn't find a better way yet!
                transmute::<*const c_void, *mut c_void>(
                    self.storage.as_ref() as *const _ as *const c_void
                )
            },
            map: Some(Self::extern_map),
        }
    }
}
```
We have an access-controlled map, handled by safe code. On top of that, there are means to create an interface a plugin can use to interact with. This mapper can also be used to map URIs to URIDs, but it does this on it's own, with it's own Rust tools. The plugin, however, has to use the interface itself and therefore is in a completely different situation, as you can see in the code of the mapping feature for the plugin.

Many things that are useful for plugins isn't useful to hosts and many things that are useful to hosts aren't useful for plugins. With this crate, we should focus on the plugin's side of things, because there will be (hopefully) many plugins that use our code, but there will be only one, maybe two hosts. These have to be developed from scratch and based on the `sys` crates nonetheless and therefore, we should not care to much about the hosts yet.

Reply:
I get that hosts and plugins have different usages of the same LV2 elements, however I did imagine features would be able to be shared, since they always seemed rather simple to implement to me (but there are just a lot of them).

I did make an URIDMap/Unmap feature usable both on the host side work in my prototype without much work ([see here](https://gitlab.com/prokopyl/lv2-rs/blob/master/lv2/src/urid/features.rs)), and at the time I didn't see any issues. What do you think about it?
(I also later imagined we could have an "host" feature in the crate that would be disabled by default, which would add the side of LV2 only hosts would use, so it would not bloat the documentation/completion for the majority of users who would only write plugins)

Also, while supporting hosts and plugins is a nice to have, the use-case I was thinking about was rather for later making an LV2 test library/harness, which would be just a host in itself to load the plugins, but where the test setup case would be able to call into the host-side of features to put them in a certain state for example, and tests the plugin reaction.

Reply:
> however I did imagine features would be able to be shared

I'm actually not so sure about that. Sure, some things have to be shared on the host side, but you can often implement it by creating an instance of something only one plugin has access to, e.g. an `Arc<Mutex>`. We should not mangle too much with it and stick with what bindgen creates from the headers. Unfortunately this means that features aren't as sharable as we would have liked.

> I did make an URIDMap/Unmap feature [...]. What do you think about it?

I see two things in your prototype:

First of all, how do you create a `URIDMap` from a raw feature reference? Let's say we have `raw_mapper` of type `&sys::LV2_URID_Map`. What do you pass to `URIDMap::new`? If you directly pass `raw_mapper`, then the internal `sys::LV2_URID_Map` will have a handle pointing to another `sys::LV2_URID_Map`. Also, this won't work because you would have to implement `URIDMapper` for it. However, passing the handle (`raw_mapper.handle`) doesn't work either, because we will never now the type of the handle.

Secondly, if you want to use it for a host, you use unsafe, dynamically dispatched code where you could use safe, statically dispatched code. It is quite hard to read out how the URID mapping is actually done, while my approach presented above shows this quite easily.

You can serve both use-cases better if you use two separate, specialized structs instead of one generalized one, and you can say that for many, maybe all features: The host is providing the interface, the plugin is using it. These are two completely different problems that should be handled by different structs.

> the use-case I was thinking about was rather for later making an LV2 test library/harness

This sounds awesome! Especially because I don't think that it already exists in the C world and since it's always hard to debug plugins, this would make thinks way easier!

Reply:
> We should not mangle too much with it and stick with what bindgen creates from the headers. Unfortunately this means that features aren't as sharable as we would have liked.

I'm not quite sure what you mean here. The LV2 headers (and so what Bindgen generates) are shared between hosts and plugins, so I'm not sure how that prevents us to do the same thing?

> 
> > I did make an URIDMap/Unmap feature [...]. What do you think about it?
> 
> I see two things in your prototype:
> 
> First of all, how do you create a `URIDMap` from a raw feature reference?

You don't. :slightly_smiling_face:

In this version, `URIDMap` *is* the raw feature and not a pointer or reference to it (since it is just a newtype around the struct from `-sys`), so if you already have pointer and want to get the `URIDMap`, you just use pointer casts (which is something only the `lv2-core` library should be doing).

The `URIDMap::new()` method actually creates a new raw feature descriptor (and the underlying `sys::LV2_URID_Map` struct), and is pretty much an host-only method (it should be feature-gated and not present by default, but the prototype didn't have that system set up). The `raw_mapper` (used as a handle) is the structure that really does the storage and mapping of URIDs (it's your underlying `HashMap` basically, but generic over the actual implementation), and again it's an host-only thing which would be separate indeed.

The `URIDMap` here has the same semantics as the raw `sys::LV2_URID_Map` (because it's really the same struct, just a newtype): it does not care about how the feature is actually implemented, it just holds the pointers to the functions it needs, plus the arbitrary handle pointer that delegates the actual implementation. 

We call `URIDMap` the feature, but technically it is just the feature descriptor, not the feature implementation itself, which is implemented by another struct (which implements the `URIDMapper` trait here, in C LV2 it is just a pointer to whatever). This is the only thing that I believe should be shared, not the implementation itself of course. :slightly_smiling_face: 

> You can serve both use-cases better if you use two separate, specialized structs instead of one generalized one, and you can say that for many, maybe all features: The host is providing the interface, the plugin is using it. These are two completely different problems that should be handled by different structs.

Indeed, and that's what is done here: `URIDMap` and `URIDUnmap` are the interface between the host and the plugin (therefore shared, like LV2 does), and they internally are just pointers to the struct actually implementing the feature (here implementing `URIDMapper`, and which contains the potential mutexes, arcs, hashmaps and more), which is an host-only problem (and probably not even provided by this crate).

The only thing I want to share here is the interface, the implementation is left to the host entirely. :slightly_smiling_face:

Reply:
I feel like this discussion is getting a bit out of hand :confused: 

Please, create a comprehensive pull request for this branch with your feature system, together with a detailed description why your version should be used instead of the current one. I think this discussion will be easier if we have the actual code side-by-side!

Reply:
I agree there ^^'

I'll work on a PR soon with the few changes I mentioned, I find it is indeed much easier to discuss with code directly!